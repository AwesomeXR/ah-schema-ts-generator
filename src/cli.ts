#!/usr/bin/env node

import * as yargs from 'yargs';
import { run } from './index';

const args = yargs
  .option('rootDir', { alias: 'r', describe: '项目根目录', type: 'string' })
  .option('inputs', { alias: 'i', describe: '输入文件 glob pattern', type: 'string', array: true })
  .option('ignores', { describe: '忽略文件 glob pattern', type: 'string', array: true })
  .option('outputDir', { alias: 'o', describe: 'ts 类型输出目录', type: 'string' })
  .demandOption(['inputs', 'outputDir'])
  .parseSync();

run({
  rootDir: args.rootDir || process.cwd(),
  inputs: args.inputs,
  ignores: args.ignores || [],
  outputDir: args.outputDir,
}).catch(err => {
  console.error(err);
  process.exit(1);
});
