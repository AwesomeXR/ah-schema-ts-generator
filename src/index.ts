export * from './createTsTypeMapFromModule';
export * from './run';
export * from './util';
