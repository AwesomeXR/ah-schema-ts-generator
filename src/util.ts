import * as Path from 'path';
import * as fs from 'fs';

// copy from https://github.com/typeorm/typeorm/blob/master/src/util/ImportUtils.ts
export async function importOrRequireFile(filePath: string): Promise<[result: any, moduleType: 'esm' | 'commonjs']> {
  const tryToImport = async (): Promise<[any, 'esm']> => {
    // `Function` is required to make sure the `import` statement wil stay `import` after
    // transpilation and won't be converted to `require`
    return [await Function('return filePath => import(filePath)')()(filePath), 'esm'];
  };

  const tryToRequire = async (): Promise<[any, 'commonjs']> => {
    return [require(filePath), 'commonjs'];
  };

  const extension = filePath.substring(filePath.lastIndexOf('.') + '.'.length);

  if (extension === 'mjs' || extension === 'mts') return tryToImport();
  else if (extension === 'cjs' || extension === 'cts') return tryToRequire();
  else if (extension === 'js' || extension === 'ts') {
    const packageJson = await getNearestPackageJson(filePath);

    if (packageJson != null) {
      const isModule = (packageJson as any)?.type === 'module';

      if (isModule) return tryToImport();
      return tryToRequire();
    }

    return tryToRequire();
  }

  return tryToRequire();
}

function getNearestPackageJson(filePath: string): Promise<object | null> {
  return new Promise(accept => {
    let currentPath = filePath;

    function searchPackageJson() {
      const nextPath = Path.dirname(currentPath);

      if (currentPath === nextPath)
        // the top of the file tree is reached
        accept(null);
      else {
        currentPath = nextPath;
        const potentialPackageJson = Path.join(currentPath, 'package.json');

        fs.stat(potentialPackageJson, (err, stats) => {
          if (err != null) searchPackageJson();
          else if (stats.isFile()) {
            fs.readFile(potentialPackageJson, 'utf8', (_err, data) => {
              if (_err != null) accept(null);
              else {
                try {
                  accept(JSON.parse(data));
                } catch (_err2) {
                  accept(null);
                }
              }
            });
          } else searchPackageJson();
        });
      }
    }

    searchPackageJson();
  });
}
