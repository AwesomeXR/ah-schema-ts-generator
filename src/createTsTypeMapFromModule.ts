import {
  isSchemaArray,
  isSchemaBoolean,
  isSchemaInteger,
  isSchemaNumber,
  isSchemaObject,
  isSchemaOneOf,
  isSchemaString,
  createTsTypeLiteral,
} from 'ah-api-type';
import { importOrRequireFile } from './util';

export type IModuleInfo = {
  tsTypeLiteral: string;
};

export const createTsTypeMapFromModule = async (filepath: string): Promise<Map<string, IModuleInfo>> => {
  const [mEntry] = await importOrRequireFile(filepath);

  const resultMap = new Map<string, IModuleInfo>();

  for (const [expName, expValue] of Object.entries(mEntry)) {
    const schema = expValue as any;
    if (!schema) continue;
    if (typeof schema !== 'object') continue;

    if (
      isSchemaArray(schema) ||
      isSchemaObject(schema) ||
      isSchemaBoolean(schema) ||
      isSchemaNumber(schema) ||
      isSchemaInteger(schema) ||
      isSchemaString(schema) ||
      isSchemaOneOf(schema)
    ) {
      const tsTypeLiteral = createTsTypeLiteral(schema);
      resultMap.set(expName, { tsTypeLiteral });
    }
  }

  return resultMap;
};
