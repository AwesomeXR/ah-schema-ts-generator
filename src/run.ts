import { createTsTypeMapFromModule, IModuleInfo } from './createTsTypeMapFromModule';
import * as Path from 'path';
import * as glob from 'glob';
import * as fs from 'fs';
import * as prettier from 'prettier';
import { Logger } from 'ah-logger';

export type IRunOption = {
  rootDir: string;
  inputs: string[];
  ignores: string[];
  outputDir: string;
};

export const run = async (opt: IRunOption) => {
  const logger = new Logger('ST');
  const { inputs, ignores, rootDir, outputDir } = opt;

  const inputSet = new Set<string>();
  for (const _p of inputs) {
    glob.sync(_p, { cwd: rootDir, absolute: true, nodir: true }).forEach(v => inputSet.add(v));
  }

  const ignoreSet = new Set<string>();
  for (const _p of ignores) {
    glob.sync(_p, { cwd: rootDir, absolute: true, nodir: true }).forEach(v => ignoreSet.add(v));
  }

  const realInputs = [...inputSet].filter(_p => !ignoreSet.has(_p) && !_p.endsWith('.d.ts'));
  const resultMap = new Map<string, Map<string, IModuleInfo>>();

  for (let i = 0; i < realInputs.length; i++) {
    const filepath = realInputs[i];

    const subMap = await createTsTypeMapFromModule(filepath);
    if (subMap.size === 0) continue;

    const relFilepath = Path.relative(rootDir, filepath);
    logger.info('[%s/%s] %s in %s', i + 1, realInputs.length, subMap.size, relFilepath);

    resultMap.set(filepath, subMap);
  }

  let content = '';
  content += 'declare interface ST {';

  for (const [filepath, subMap] of resultMap) {
    const relFilepath = Path.relative(rootDir, filepath);

    content += `// auto generate from ${relFilepath}\n`;
    for (const [key, info] of subMap) {
      content += `${key}: ${info.tsTypeLiteral};\n`;
    }
  }

  content += '}';

  // pretty
  const prettierOpt = await prettier.resolveConfig(rootDir);
  content = await prettier.format(content, {
    parser: 'typescript',
    trailingComma: 'es5',
    tabWidth: 2,
    semi: true,
    singleQuote: true,
    jsxSingleQuote: true,
    printWidth: 120,
    bracketSpacing: true,
    arrowParens: 'avoid',
    ...prettierOpt,
  });

  fs.writeFileSync(Path.join(outputDir, 'declare.d.ts'), content, 'utf-8');
};
