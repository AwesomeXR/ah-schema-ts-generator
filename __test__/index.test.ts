import { createTsTypeMapFromModule } from '../src';
import * as Path from 'path';

describe('createTsTypeMapFromModule', () => {
  it.skip('test', async () => {
    const map = await createTsTypeMapFromModule(Path.resolve(__dirname, './fixture.ts'));
    expect(map).toMatchSnapshot();
  });
});
