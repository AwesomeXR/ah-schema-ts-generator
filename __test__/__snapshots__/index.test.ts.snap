// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createTsTypeMapFromModule test 1`] = `
Map {
  "IGetDetailReq" => {
    "tsTypeLiteral": "{
totalCheckInCnt?: number;
readUVCnt?: number;

/** (persons_desc) */
persons?: Array<{
id: number;
cnName?: string;
receiveFlowerTotalCount?: number;
receiveFlowerUserCount?: number;
pavilionCategoryId?: number;
cover?: string;
deedDesc?: string;
isDead?: any
}>
}",
  },
}
`;
